File: src/dsa/problems/HouseRobber.java
package dsa.problems;
//LC# 198
/**
 * the last decision made by the robber could be either to rob a house or not
 * F(n) = max.no.of houses you can rob from first n houses =
 * max{
 *      f(n-1) ->(to not rob this house),
 *      f(n-2) + nums[n] -> rob this house and accrue the money
 * }
 */
public class HouseRobber {
    public int solve(int []a) {
        int[]dp = new int[a.length];
        //base case, if there were no houses, total value a robber would get be 0
        dp[0] = 0;
        dp[1] = a[1];
        dp[2] = Math.max(a[1], a[2]);
        for (int i = 3; i < a.length; i++) {
            dp[i] = Math.max(dp[i -1], dp[i -2] + a[i]);
        }
        return dp[a.length - 1];
    }
    //LC# 213
    /**
     * If N houses are arranged in a circular manner, we need to somehow break the circle, then we arrive at following cases
     * Case 1: House #0 is robbed
     * we need to find an optimal solution for houses between 1 to N - 2. We cannot rob house # N -1,
     * for this we can use the "straight line" soltn(prev problem)
     *
     * Case 2:house #0 is NOT robbed
     * we need to find an optimal solution for houses between 1 to N - 1, for this we can use the "straight line" soltn(prev problem)
     *Finally we return the max value from (case 1 soltn, case 2 soltn)
     */
}

File: src/dsa/python/dp/house_robber.py
#link: https://leetcode.com/problems/house-robber/description/
from typing import List


def solve_2(nums: List[int]):
    N = len(nums)
    if N == 0:
        return 0
    if N == 1:
        return nums[0]
    dp=[0]*N
    dp[0] = nums[0]
    dp[1] = max(nums[0], nums[1])

    for i in range(2, N):
        #the last house say i could either be robbed or not robbed.
        # if it is robbed, then previous house cannot be robbed, so look for solutions from 0 to i -2
        # if it is NOT robbed, then look for solutions from 0 to i - 1
        dp[i] = max(dp[i - 1], nums[i] + dp[i - 2])
    return dp[N - 1]


def solve(nums: List[int]):
    n = len(nums)
    if n == 0:
        return 0
    if n == 1:
        return nums[0]

    ans = [0]*n #ans[i]=max amount of money you can rob from houses 0...i - 1
    ans[0] = nums[0]
    ans[1] = max(nums[0], nums[1])
    if n == 2:
        return ans[1]

    ans[2] = ans[0] + nums[2]
    for i in range(3, n):
        # we cannot take the house just before i, so we can either take i - 2, or i - 3 house
        ans[i] = max(nums[i] + ans[i - 2], nums[i] + ans[i - 3])
    return max(ans[n - 1], ans[n - 2])


def main():
    assert solve([1, 2, 3, 1]) == 4
    assert solve_2([1, 2, 3, 1]) == 4
    assert solve([2, 7, 9, 3, 1]) == 12
    assert solve([1]) == 1
    assert solve([0,0]) == 0
    assert solve_2([0,0]) == 0


if __name__ == '__main__':
    main()
